### ISSUE TITLE
Left join should filter first if predicate on join key (`WARN Skipping record due to null join key or value.`)

### ISSUE THREAD

**Issue by rmoff** [confluentinc/ksql#7274]
I'm not sure if this is a bug, or an optimisation … :) 

When joining on a function my ksqlDB server log is spammed with lots of `WARN Skipping record due to null join key or value.`. I was troubleshooting another data issue and this turned out to be a big red herring in the end, but also suggests that there could be more optimal execution. Thus: 

If you do a stream-table join and the key field in the stream is a function that returns null, this record shouldn't be filtered out up front, but isn't, even with an explicit predicate. 

This is on ksqlDB 0.15. 

* Set up the test data

```sql
CREATE TABLE LOOKUP (PK VARCHAR PRIMARY KEY, NAME VARCHAR) WITH (VALUE_FORMAT='AVRO', PARTITIONS=1, KAFKA_TOPIC='LOOKUP');
INSERT INTO LOOKUP VALUES ('A42','FOO');

CREATE STREAM EVENTS (E VARCHAR) WITH (VALUE_FORMAT='AVRO', KAFKA_TOPIC='EVENTS', PARTITIONS=1);
INSERT INTO EVENTS VALUES ('{"v1":"a","fk":"A42"}');
INSERT INTO EVENTS VALUES ('{"v1":"b","fk":"X"}');
INSERT INTO EVENTS VALUES ('{"v1":"C"}');
INSERT INTO EVENTS VALUES ('{"v1":"D","fk":"A42"}');

SET 'auto.offset.reset' = 'earliest';
ksql> SELECT * FROM LOOKUP EMIT CHANGES;
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|PK                                                                                          |NAME                                                                                        |
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|A42                                                                                         |FOO                                                                                         |
^CQuery terminated

ksql> SELECT * FROM EVENTS EMIT CHANGES;
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|E                                                                                                                                                                                         |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|{"v1":"a","fk":"A42"}                                                                                                                                                                     |
|{"v1":"b","fk":"X"}                                                                                                                                                                       |
|{"v1":"C"}                                                                                                                                                                                |
|{"v1":"D","fk":"A42"}                                                                                                                                                                     |
^CQuery terminated
```

* Examine the join output

```sql
ksql> SELECT EXTRACTJSONFIELD(E,'$.v1') AS V1, NAME FROM EVENTS EV LEFT JOIN LOOKUP L ON EXTRACTJSONFIELD(E,'$.fk') = L.PK EMIT CHANGES;
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|V1                                                                                          |NAME                                                                                        |
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|a                                                                                           |FOO                                                                                         |
|b                                                                                           |null                                                                                        |
|D                                                                                           |FOO                                                                                         |
```

* Create a joined stream.

```sql
ksql> CREATE STREAM JOINED AS SELECT EXTRACTJSONFIELD(E,'$.v1') AS V1, NAME FROM EVENTS EV LEFT JOIN LOOKUP L ON EXTRACTJSONFIELD(E,'$.fk') = L.PK PARTITION BY EXTRACTJSONFIELD(E,'$.v1');
```

Results are as expected: 

```
ksql> SELECT * FROM JOINED EMIT CHANGES;
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|V1                                                                                                                                                                                          |NAME                                                                                                                                                                                        |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
|a                                                                                                                                                                                           |FOO                                                                                                                                                                                         |
|b                                                                                                                                                                                           |null                                                                                                                                                                                        |
|C                                                                                                                                                                                           |FOO                                                                                                                                                                                         |
```

Here is the issue - the resulting null key records in the repartition topic which are quite rightly logged: 

```
[2021-03-23 10:55:20,619] WARN Skipping record due to null join key or value. key=[null] value=[[ '{"v1":"C"}' | 1616496854917L | null ]] topic=[_confluent-ksql-confluent_rmoff_02query_CSAS_JOINED_41-Join-repartition] partition=[0] offset=[2] (org.apache.kafka.streams.kstream.internals.KStreamKTableJoin:71)
```

With lots of data this gets pretty noisy in the log. Even if you try to filter them out, the `WHERE` seems not to be applied before the repartition: 

```
ksql> CREATE STREAM JOINED2 AS SELECT EXTRACTJSONFIELD(E,'$.v1') AS V1, NAME FROM EVENTS EV LEFT JOIN LOOKUP L ON EXTRACTJSONFIELD(E,'$.fk') = L.PK WHERE EXTRACTJSONFIELD(E,'$.fk') IS NOT NULL PARTITION BY EXTRACTJSONFIELD(E,'$.v1');

 Message
---------------------------------------
 Created query with ID CSAS_JOINED2_43
---------------------------------------

ksql> SELECT * FROM JOINED2 EMIT CHANGES;
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|V1                                                                                          |NAME                                                                                        |
+--------------------------------------------------------------------------------------------+--------------------------------------------------------------------------------------------+
|a                                                                                           |FOO                                                                                         |
|b                                                                                           |null                                                                                        |
|D                                                                                           |FOO                                                                                         |
```

Log entry: 

```
[2021-03-23 10:57:06,742] WARN Skipping record due to null join key or value. key=[null] value=[[ '{"v1":"C"}' | 1616496854917L | null ]] topic=[_confluent-ksql-confluent_rmoff_02query_CSAS_JOINED2_43-Join-repartition] partition=[0] offset=[2] (org.apache.kafka.streams.kstream.internals.KStreamKTableJoin:71)
```

**Comment by vcrfxia** [comment]
Hi @rmoff , your report indeed matches what I'm seeing based on the generated Kafka Streams topology for a similar join. Looks like the where filter is applied after the join, though ideally it'd be applied before. 

Here are the sample statements and topology, in case anyone else wants to play around with this in the [visualizer](https://zz85.github.io/kafka-streams-viz/):
```
      "statements": [
        "CREATE TABLE T (ID STRING PRIMARY KEY, VAL STRING) WITH (kafka_topic='t', key_format='KAFKA', value_format='JSON');",
        "CREATE STREAM S (ID STRING KEY, FOO INT) WITH (kafka_topic='s', key_format='DELIMITED', value_format='JSON');",
        "CREATE STREAM OUTPUT AS SELECT T.ID, S.ID, VAL FROM S JOIN T ON EXTRACTJSONFIELD(S.ID,'$.fk') = T.ID WHERE EXTRACTJSONFIELD(S.ID,'$.fk') IS NOT NULL;"
      ],
```
```
Topologies:
   Sub-topology: 0
    Source: Join-repartition-source (topics: [Join-repartition])
      --> Join
    Source: KSTREAM-SOURCE-0000000001 (topics: [t])
      --> KTABLE-SOURCE-0000000002
    Processor: Join (stores: [KafkaTopic_Right-Reduce])
      --> WhereFilter
      <-- Join-repartition-source
    Processor: KTABLE-SOURCE-0000000002 (stores: [])
      --> KTABLE-MAPVALUES-0000000003
      <-- KSTREAM-SOURCE-0000000001
    Processor: KTABLE-MAPVALUES-0000000003 (stores: [KafkaTopic_Right-Reduce])
      --> KTABLE-TRANSFORMVALUES-0000000004
      <-- KTABLE-SOURCE-0000000002
    Processor: WhereFilter (stores: [])
      --> Project
      <-- Join
    Processor: KTABLE-TRANSFORMVALUES-0000000004 (stores: [])
      --> PrependAliasRight
      <-- KTABLE-MAPVALUES-0000000003
    Processor: Project (stores: [])
      --> KSTREAM-SINK-0000000016
      <-- WhereFilter
    Sink: KSTREAM-SINK-0000000016 (topic: OUTPUT)
      <-- Project
    Processor: PrependAliasRight (stores: [])
      --> none
      <-- KTABLE-TRANSFORMVALUES-0000000004

  Sub-topology: 1
    Source: KSTREAM-SOURCE-0000000006 (topics: [s])
      --> KSTREAM-TRANSFORMVALUES-0000000007
    Processor: KSTREAM-TRANSFORMVALUES-0000000007 (stores: [])
      --> LeftSourceKeyed-SelectKey
      <-- KSTREAM-SOURCE-0000000006
    Processor: LeftSourceKeyed-SelectKey (stores: [])
      --> PrependAliasLeft
      <-- KSTREAM-TRANSFORMVALUES-0000000007
    Processor: PrependAliasLeft (stores: [])
      --> Join-repartition-filter
      <-- LeftSourceKeyed-SelectKey
    Processor: Join-repartition-filter (stores: [])
      --> Join-repartition-sink
      <-- PrependAliasLeft
    Sink: Join-repartition-sink (topic: Join-repartition)
      <-- Join-repartition-filter
```

**Comment by mjsax** [comment]
Seem there are two thing here:

First:

> If you do a stream-table join and the key field in the stream is a function that returns null, this record shouldn't be filtered out up front,

Agreed. This is already tracked: https://issues.apache.org/jira/browse/KAFKA-12317

Second:

> With lots of data this gets pretty noisy in the log. Even if you try to filter them out, the WHERE seems not to be applied before the repartition:

> I'm not sure if this is a bug, or an optimisation …

It's definitely an optimization. And as a matter of fact, even if there is an optimizer, it may decide to place a filter before or after the join. There is no guarantee whatsoever, in which order operators will be executed and it's to the sole discretion to the optimizer how to rewrite a plan.

Thus, even if most likely a filter push down would happen (if we would have an optimizer), you should never write a program that relies on this optimization to be applied.

Bottom line: I think we can just close this ticket. The first issue must be address in Kafka Streams and there is a ticket. And the second "issue" is no issue from my POV.

**Comment by vcrfxia** [comment]
Turns out we already had a ticket for the join/filter order optimization: https://github.com/confluentinc/ksql/issues/4748
